### PR TITLE
Adds revision ranges

### PR DESCRIPTION
**Pull request by boehmseb**
This PR introduces **revision ranges**. A revision range represents a sub-graph of a git-history by various means. Such ranges can be used to customize project behavior on a per-revision basis without the need to list each revision separately.

As an example, I included a decorator that allows to mark revisions in a project that should be ignored during experiments as _blocked_.

Open TODOs:
- [ ] Revision ranges need access to a git repo to enumerate included revisions or to make containment-checks like `range.contains(commit)` (@simbuerg help needed)
- [ ] revision block example project
- [x] Tests

### CODE REVIEW

File: benchbuild/utils/revision_ranges.py [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+            self.__id_start, self.__id_end, git)
+
+    def __iter__(self) -> tp.Iterator[str]:
+        if self.__revision_list is None:
+            raise AssertionError
+        return self.__revision_list.__iter__()
+
+
+class CommitState(IntFlag):
+    BOT = 0
```
Comment (vulder) [inline benchbuild/utils/revision_ranges.py:134]: Are you using BOT as the bottom elem or is this something else?

File: benchbuild/utils/revision_ranges.py [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+                for b_item in b_entry:
+                    if b_item.startswith(rev_id):
+                        return True, b_entry.comment
+            return False, None
+
+        cls.blocked_revisions_initialized = False
+        cls.is_blocked_revision = is_blocked_revision_impl
+        return cls
+
+    return revision_blocker_decorator
```
Comment (vulder) [inline benchbuild/utils/revision_ranges.py:276]: Do we want this on BB level? Not sure if blocking revisions is useful for BB.
@simbuerg

File: benchbuild/utils/revision_ranges.py (hunk 1 above)
Comment (boehmseb) [reply benchbuild/utils/revision_ranges.py:134]: BOT means bottom here. I can rename this.

**Comment by simbuerg** [comment]
The abstraction benchbuild.utils.cmd is used for configurable command aliasing and env adjustment.
You can use it as a drop-in for plumbum commands just fine and gain control over the command environment
via benchbuild's settings.
This is important for situations where you require locally installed tools with custom LD_LIBRARY_PATH entries,
or other environment variables. Without this, you need to make sure to setup the environment correctly yourself.

However, you are using the local machine of plumbum directly (...import local).
There is no equivalent in benchbuild, so it is fine.

Am Wednesday, den 06.05.2020, 11:03 -0700 schrieb Florian Sattler:
> @vulder commented on this pull request.
> 
> 
> 
> > @@ -0,0 +1,276 @@
> +"""
> +Revision ranges model a sub-graph of the git history by various means and can be
> +used to modify project behaviour on a per-revision basis.
> +"""
> +
> +import abc
> +import typing as tp
> +from enum import IntFlag
> +from pathlib import Path
> +
> +import pygit2
> +from plumbum import local
> +from plumbum.machines import LocalCommand
> +
> +local_git = local["git"]
> 
> BB has it's own abstraction around `plumbum.utils.cmd` I'm not sure if we should use it here.
> @simbuerg what do you think?
> 
> > +        # cache for commit hashes
> +        self.__revision_list: tp.Optional[tp.List[str]] = None
> +
> +    def init_cache(self, repo_path: Path) -> None:
> +        git = _get_git_for_path(repo_path)
> +        self.__revision_list = _get_all_revisions_between(
> +            self.__id_start, self.__id_end, git)
> +
> +    def __iter__(self) -> tp.Iterator[str]:
> +        if self.__revision_list is None:
> +            raise AssertionError
> +        return self.__revision_list.__iter__()
> +
> +
> +class CommitState(IntFlag):
> +    BOT = 0
> 
> Are you using BOT as the bottom elem or is this something else?
> 
> > +def block_revisions(blocks: tp.List[AbstractRevisionRange]) -> tp.Any:
> +    """
> +    Decorator for project classes for blacklisting/blocking revisions.
> +
> +    ATTENTION: This decorator depends on things introduced by the
> +    @with_git decorator and therefore must be used above that decorator.
> +
> +    This adds a new static method ``is_blocked_revision`` that checks
> +    whether a given revision id is marked as blocked.
> +
> +    Args:
> +        blocks: A list of :class:`AbstractRevisionRange` s.
> +    """
> +    def revision_blocker_decorator(cls: tp.Any) -> tp.Any:
> +        def is_blocked_revision_impl(
> +                rev_id: str) -> tp.Tuple[bool, tp.Optional[str]]:
> +            """
> +            Checks whether a revision is blocked or not. Also returns the
> +            reason for the block if available.
> +            """
> +            # trigger caching for BlockedRevisionRanges
> +            if not cls.blocked_revisions_initialized:
> +                cls.blocked_revisions_initialized = True
> +                for block in blocks:
> +                    # TODO: how to access local repository?
> +                    block.init_cache(cls.repo_path())
> +
> +            for b_entry in blocks:
> +                for b_item in b_entry:
> +                    if b_item.startswith(rev_id):
> +                        return True, b_entry.comment
> +            return False, None
> +
> +        cls.blocked_revisions_initialized = False
> +        cls.is_blocked_revision = is_blocked_revision_impl
> +        return cls
> +
> +    return revision_blocker_decorator
> 
> Do we want this on BB level? Not sure if blocking revisions is useful for BB.
> @simbuerg 
>

**Comment by simbuerg** [comment]
Why would you rename it? Your Code :-)

Am Wednesday, den 06.05.2020, 11:07 -0700 schrieb Sebastian Böhm:
> @boehmseb commented on this pull request.
> 
> 
> 
> > +        # cache for commit hashes
> +        self.__revision_list: tp.Optional[tp.List[str]] = None
> +
> +    def init_cache(self, repo_path: Path) -> None:
> +        git = _get_git_for_path(repo_path)
> +        self.__revision_list = _get_all_revisions_between(
> +            self.__id_start, self.__id_end, git)
> +
> +    def __iter__(self) -> tp.Iterator[str]:
> +        if self.__revision_list is None:
> +            raise AssertionError
> +        return self.__revision_list.__iter__()
> +
> +
> +class CommitState(IntFlag):
> +    BOT = 0
> 
> BOT means bottom here. I can rename this.
>

File: benchbuild/utils/revision_ranges.py (hunk 1 above)
Comment (vulder) [reply benchbuild/utils/revision_ranges.py:134]: I thought so :)

File: benchbuild/utils/revision_ranges.py [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+import abc
+import typing as tp
+from enum import IntFlag
+from pathlib import Path
+
+import pygit2
+from plumbum import local
+from plumbum.machines import LocalCommand
+
+local_git = local["git"]
```
[the remark this answers is not in the document]
Comment (simbuerg) [reply benchbuild/utils/revision_ranges.py:15]: The abstraction is useful for aliasing and environment control for the imported commands. Do you need to use local directly? If not, you can just drop this one in:

`from benchbuild.utils.cmd import git as local_git`

Or don't rename it, if the local_ prefix is arbitrary.

File: benchbuild/utils/revision_ranges.py (hunk 2 above)
Comment (simbuerg) [reply benchbuild/utils/revision_ranges.py:276]: That is totally under the control of a project. We can always rename/generalize it later.

**Comment by simbuerg** [comment]
That looks simple enough to integrate. I would suggest we add it in and integrate/refactor anything away afterwards. It needs transformation inside the f/environment pull request anyway.

Do you have a few simple unit tests for me? That would help when I merge it to the new f/environments pull request.

File: benchbuild/utils/revision_ranges.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (boehmseb) [reply benchbuild/utils/revision_ranges.py:15]: The name is arbitrary but I would like to have one, so I can name parameters `git` without running into conflicts.

**Comment by boehmseb** [comment]
> Do you have a few simple unit tests for me? That would help when I merge it to the new f/environments pull request.
I added some basic unit tests for the range-logic, but anything more would either require a real git repository or some extensive mocking.

**Comment by simbuerg** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 6
           

Complexity increasing per file
==============================
- tests/test_revision_ranges.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/polyjit/benchbuild_2/pullRequest?prid=5481069&bid=17962371)

File: tests/test_revision_ranges.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
+import unittest
```
Comment (simbuerg) [inline tests/test_revision_ranges.py:1]: Codacy Issue found: [Missing module docstring](https://app.codacy.com/manual/polyjit/benchbuild_2/pullRequest?prid=5481069)

File: tests/test_revision_ranges.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+from pathlib import Path
+import unittest.mock as mock
+
+import benchbuild.utils.revision_ranges as ranges
+
+
+class TestRevisionRanges(unittest.TestCase):
```
Comment (simbuerg) [inline tests/test_revision_ranges.py:8]: Codacy Issue found: [Missing class docstring](https://app.codacy.com/manual/polyjit/benchbuild_2/pullRequest?prid=5481069)

File: tests/test_revision_ranges.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+        commit_head.parents = [commit_good, commit_y]
+        commit_good.parents = [commit_r, commit_s]
+        commit_bad.parents = [commit_x]
+        commit_r.parents = [commit_x, commit_bad]
+        commit_s.parents = [commit_bad]
+        commit_x.parents = []
+        commit_y.parents = []
+
+        result = ranges._find_blocked_commits(commit_head, [commit_good],
+                                       [commit_bad])
```
Comment (simbuerg) [inline tests/test_revision_ranges.py:48]: Codacy Issue found: [Wrong continued indentation (add 7 spaces).](https://app.codacy.com/manual/polyjit/benchbuild_2/pullRequest?prid=5481069)

File: tests/test_revision_ranges.py [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+        """
+        commit_head.parents = [commit_good, commit_y]
+        commit_good.parents = [commit_r, commit_s]
+        commit_bad.parents = [commit_x]
+        commit_r.parents = [commit_x, commit_bad]
+        commit_s.parents = [commit_bad]
+        commit_x.parents = []
+        commit_y.parents = []
+
+        result = ranges._find_blocked_commits(commit_head, [commit_good],
```
Comment (simbuerg) [inline tests/test_revision_ranges.py:47]: Codacy Issue found: [Access to a protected member _find_blocked_commits of a client class](https://app.codacy.com/manual/polyjit/benchbuild_2/pullRequest?prid=5481069)

File: tests/test_revision_ranges.py [hunk 8]
```diff
@@ -0,0 +1,51 @@
+import unittest
+from pathlib import Path
+import unittest.mock as mock
```
Comment (simbuerg) [inline tests/test_revision_ranges.py:3]: Codacy Issue found: [Imports from package unittest are not grouped](https://app.codacy.com/manual/polyjit/benchbuild_2/pullRequest?prid=5481069)

File: tests/test_revision_ranges.py [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+            self.assertIn(commit, revision_range)
+
+    @mock.patch("pygit2.Commit")
+    @mock.patch("pygit2.Commit")
+    @mock.patch("pygit2.Commit")
+    @mock.patch("pygit2.Commit")
+    @mock.patch("pygit2.Commit")
+    @mock.patch("pygit2.Commit")
+    @mock.patch("pygit2.Commit")
+    def test_find_blocked_commits(self, commit_head, commit_good, commit_bad,
```
Comment (simbuerg) [inline tests/test_revision_ranges.py:30]: Codacy Issue found: [Too many arguments (8/5)](https://app.codacy.com/manual/polyjit/benchbuild_2/pullRequest?prid=5481069)

**Comment by simbuerg** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- tests/test_revision_ranges.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/polyjit/benchbuild_2/pullRequest?prid=5481069&bid=17962371)

File: tests/test_revision_ranges.py [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+        """
+        commit_head.parents = [commit_good, commit_y]
+        commit_good.parents = [commit_r, commit_s]
+        commit_bad.parents = [commit_x]
+        commit_r.parents = [commit_x, commit_bad]
+        commit_s.parents = [commit_bad]
+        commit_x.parents = []
+        commit_y.parents = []
+
+        result = ranges._find_blocked_commits(commit_head, [commit_good],
```
Comment (simbuerg) [inline tests/test_revision_ranges.py:50]: Codacy Issue found: [Access to a protected member _find_blocked_commits of a client class](https://app.codacy.com/manual/polyjit/benchbuild_2/pullRequest?prid=5481069)

File: tests/test_revision_ranges.py [hunk 11]
```diff
@@ -0,0 +1,54 @@
+"""Test the revision_ranges module."""
+import unittest
+from pathlib import Path
+from unittest import mock
```
Comment (simbuerg) [inline tests/test_revision_ranges.py:4]: Codacy Issue found: [Imports from package unittest are not grouped](https://app.codacy.com/manual/polyjit/benchbuild_2/pullRequest?prid=5481069)

File: tests/test_revision_ranges.py [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+            self.assertIn(commit, revision_range)
+
+    @mock.patch("pygit2.Commit")
+    @mock.patch("pygit2.Commit")
+    @mock.patch("pygit2.Commit")
+    @mock.patch("pygit2.Commit")
+    @mock.patch("pygit2.Commit")
+    @mock.patch("pygit2.Commit")
+    @mock.patch("pygit2.Commit")
+    def test_find_blocked_commits(self, commit_head, commit_good, commit_bad,
```
Comment (simbuerg) [inline tests/test_revision_ranges.py:33]: Codacy Issue found: [Too many arguments (8/5)](https://app.codacy.com/manual/polyjit/benchbuild_2/pullRequest?prid=5481069)

**Comment by simbuerg** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- tests/test_revision_ranges.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/polyjit/benchbuild_2/pullRequest?prid=5481069&bid=17962371)

File: tests/test_revision_ranges.py [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+        """
+        commit_head.parents = [commit_good, commit_y]
+        commit_good.parents = [commit_r, commit_s]
+        commit_bad.parents = [commit_x]
+        commit_r.parents = [commit_x, commit_bad]
+        commit_s.parents = [commit_bad]
+        commit_x.parents = []
+        commit_y.parents = []
+
+        result = ranges._find_blocked_commits(commit_head, [commit_good],
```
Comment (simbuerg) [inline tests/test_revision_ranges.py:49]: Codacy Issue found: [Access to a protected member _find_blocked_commits of a client class](https://app.codacy.com/manual/polyjit/benchbuild_2/pullRequest?prid=5481069)

File: tests/test_revision_ranges.py [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+            self.assertIn(commit, revision_range)
+
+    @mock.patch("pygit2.Commit")
+    @mock.patch("pygit2.Commit")
+    @mock.patch("pygit2.Commit")
+    @mock.patch("pygit2.Commit")
+    @mock.patch("pygit2.Commit")
+    @mock.patch("pygit2.Commit")
+    @mock.patch("pygit2.Commit")
+    def test_find_blocked_commits(self, commit_head, commit_good, commit_bad,
```
Comment (simbuerg) [inline tests/test_revision_ranges.py:32]: Codacy Issue found: [Too many arguments (8/5)](https://app.codacy.com/manual/polyjit/benchbuild_2/pullRequest?prid=5481069)

**Comment by simbuerg** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 7
           

Complexity increasing per file
==============================
- tests/test_revision_ranges.py  2
- benchbuild/utils/revision_ranges.py  15
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/polyjit/benchbuild_2/pullRequest?prid=5481069&bid=17962371)

**Pull request merged**